src/throttler/remote_throttler.js

```diff
@@ -0,0 +1,180 @@
+
+  _refreshIntervalMs: number;
+  _host: string;
+  _port: number;
+
+  _uuid: string;
+  _credits: CreditsPerOperation;
+
+  _initialDelayTimeoutHandle: any;
+  _refreshIntervalHandle: any;
+  _onCreditsUpdate: ?Function;
+
+  /**
+   * Creates a RemoteThrottler that fetches credits remotely from jaeger-agent.
+   *
+   * @param {string} [serviceName] - name of the current service / application, same as given to Tracer
+   * @param {object} [options] - optional settings
+   * @param {object} [options.logger] - optional logger, see _flow/logger.js
+   * @param {object} [options.metrics] - instance of Metrics object
+   * @param {number} [options.refreshIntervalMs] - interval in milliseconds that determines how often credits
+   * are fetched (0 to not refresh, NOT RECOMMENDED!)
+   * @param {number} [options.initialDelayMs] - interval in milliseconds that determines how soon after initialization
+   * credits are first fetched (0 to not refresh, NOT RECOMMENDED!).
+   * @param {string} [options.host] - host for jaeger-agent, defaults to 'localhost'
+   * @param {number} [options.port] - port for jaeger-agent for /credits endpoint
+   * @param {function} [options.onCreditsUpdate] - callback function once credits are updated. Used for testing.
+   */
+  constructor(serviceName: string, options: any = {}) {
+    this._serviceName = serviceName;
+    this._logger = options.logger || new NullLogger();
+    this._metrics = options.metrics || new Metrics(new NoopMetricFactory());
+    this._refreshIntervalMs = options.refreshIntervalMs || DEFAULT_REFRESH_INTERVAL_MS;
+    this._host = options.host || DEFAULT_THROTTLER_HOST;
+    this._port = options.port || DEFAULT_THROTTLER_PORT;
+
+    this._credits = Object.create(null);
+    this._onCreditsUpdate = options.onCreditsUpdate;
+
+    this._initialDelayTimeoutHandle = setTimeout(
+      this._afterInitialDelay.bind(this),
+      options.initialDelayMs || DEFAULT_INITIAL_DELAY_MS
+    );
+  }
+
+  _afterInitialDelay(): void {
+    this._refreshCredits();
+    this._refreshIntervalHandle = setInterval(this._refreshCredits.bind(this), this._refreshIntervalMs);
+    this._initialDelayTimeoutHandle = null;
+  }
+
+  setProcess(process: Process): void {
+    this._uuid = process.uuid || '';
+  }
+
+  isAllowed(operation: string): boolean {
+    if (operation in this._credits) {
+      return this._isAllowed(operation);
+    }
+    this._credits[operation] = 0;
+    // If seen for the first time, async fetch credits
+    this._refreshCredits();
+    return false;
+  }
+
+  _isAllowed(operation: string): boolean {
+    const credits = this._credits[operation] || 0;
+    if (credits < UNIT_CREDIT) {
+      return false;
+    }
+    this._credits[operation] = credits - UNIT_CREDIT;
+    return true;
+  }
+
+  _refreshCredits() {
+    if (!this._uuid) {
+      this._logger.error(`UUID must be set to fetch credits`);
+      return;
+    }
+    const keys = Object.keys(this._credits);
+    if (keys.length == 0) {
```

Comment:
I recommend using `===` instead of `==` for all cases except comparisons with `null` or `undefined`.

Ref: [Loose equality using ==](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#Loose_equality_using)